libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp

```diff
@@ -217,143 +239,198 @@ void AP_Proximity_RPLidarA2::init_sectors()
+    _uart->write(tx_buffer, 2);
+    Debug(3, "sent reset request");
+    _reset = true;           // Lock the lidar for the time of reseting
+    _last_request_ms =  AP_HAL::millis();
+    _rp_state = rp_reset;
+}
+
 // set Lidar into SCAN mode
-void AP_Proximity_RPLidarA2::set_scan_mode()
+void AP_Proximity_RPLidarA2::send_request_for_scan()
 {
     if (_uart == nullptr) {
         return;
     }
     uint8_t tx_buffer[2] = {RPLIDAR_PREAMBLE, RPLIDAR_CMD_SCAN};
     _uart->write(tx_buffer, 2);
     _last_request_ms = AP_HAL::millis();
-    Debug(1, "LIDAR SCAN MODE ACTIVATED");
+    Debug(DBG_GCS|DBG_UART, "activated scan mode");
+    _rp_state = rp_responding;
+}
+
+// set Lidar into FORCE SCAN mode (scanning independently of the real speed)
+void AP_Proximity_RPLidarA2::send_request_for_force_scan()
+{
+    if (_uart == nullptr) {
+        return;
+    }
+    uint8_t tx_buffer[2] = {RPLIDAR_PREAMBLE, RPLIDAR_CMD_FORCE_SCAN};
+    _uart->write(tx_buffer, 2);
+    _last_request_ms = AP_HAL::millis();
+    Debug(DBG_GCS|DBG_UART, "activated scan mode");
     _rp_state = rp_responding;
 }
 
 // send request for sensor health
-void AP_Proximity_RPLidarA2::send_request_for_health()                                    //not called yet
+void AP_Proximity_RPLidarA2::send_request_for_health()
 {
     if (_uart == nullptr) {
         return;
     }
     uint8_t tx_buffer[2] = {RPLIDAR_PREAMBLE, RPLIDAR_CMD_GET_DEVICE_HEALTH};
     _uart->write(tx_buffer, 2);
     _last_request_ms = AP_HAL::millis();
-    _rp_state = rp_health;
+    Debug(DBG_GCS|DBG_UART, "sent health request");
+    _rp_state = rp_responding;
+}
+
+// send request for sensor sampling rate
+void AP_Proximity_RPLidarA2::send_request_for_rate()
+{
+    if (_uart == nullptr) {
+        return;
+    }
+    uint8_t tx_buffer[2] = {RPLIDAR_PREAMBLE, RPLIDAR_CMD_GET_DEVICE_RATE};
+    _uart->write(tx_buffer, 2);
+    _last_request_ms = AP_HAL::millis();
+    Debug(DBG_GCS|DBG_UART, "sent rate request");
+    _rp_state = rp_responding;
 }
 
 void AP_Proximity_RPLidarA2::get_readings()
 {
     if (_uart == nullptr) {
         return;
     }
-    Debug(2, "             CURRENT STATE: %d ", _rp_state);
+
+    static uint8_t sync_buff[2] = {0};
+
     uint32_t nbytes = _uart->available();
 
-    while (nbytes-- > 0) {
+    Debug(0, "current state: %d ", _rp_state);
+    Debug(0, "nbytes: %d", nbytes);
+    if(nbytes == RX_BUFFER_SIZE) {
+        Debug(DBG_GCS|DBG_UART, "UART BUFF FULL (%d)!", nbytes);
+    }
 
+    /* After reset: wait for the welcome message to finish and take the next action.
+     * The message is not documented anywhere and its length varies in different RP models,
+     * thus we don't want to base on its content and use a predefined time instead.
+     * The time is counted since receiving the first character of a welcome message. */
+    if (_rp_state == rp_reset)
+    {
+        // Do if the welcome message is being transmitted
```

Comment:
A 3 M 1 can issue commands after 2 ms have elapsed after reset.

Reply:
The code starts to listen for the welcome message after RESET_RPA2_WAIT_MS delay since issuing a RESET request. This is to account for some "leftover" data if the request is issued during the SCAN mode (it takes some time before lidar reacts).